File: src/system/installer.ts
/**
 * src/system/installer.ts — Auto-Installer & Environment Bootstrapper Engine
 */

import fs from 'node:fs';
import path from 'node:path';
import os from 'node:os';
import { execSync } from 'node:child_process';
import type { InstallOptions, InstallReport, InstallTarget, HostPlatform } from './types.ts';

export class AutoInstaller {
  private projectDir: string;
  private homeDir: string;

  constructor(projectDir: string = '.') {
    this.projectDir = path.resolve(projectDir);
    this.homeDir = os.homedir();
  }

  private checkBinary(cmd: string): { available: boolean; version?: string } {
    try {
      const out = execSync(`${cmd} --version`, { stdio: 'pipe', encoding: 'utf-8' }).trim();
      return { available: true, version: out.split('\n')[0] };
    } catch {
      return { available: false };
    }
  }

  public getHostTargets(): InstallTarget[] {
    const targets: InstallTarget[] = [
      {
        name: 'Claude Code Skills',
        platform: 'claude',
        path: path.join(this.homeDir, '.claude', 'skills', 'agentic-workflow'),
        installed: fs.existsSync(path.join(this.homeDir, '.claude', 'skills', 'agentic-workflow'))
      },
      {
        name: 'Gemini CLI / Antigravity Skills',
        platform: 'gemini',
        path: path.join(this.homeDir, '.gemini', 'config', 'skills', 'agentic-workflow'),
        installed: fs.existsSync(path.join(this.homeDir, '.gemini', 'config', 'skills', 'agentic-workflow'))
      },
      {
        name: 'Cursor IDE Skills',
        platform: 'cursor',
        path: path.join(this.homeDir, '.cursor', 'skills', 'agentic-workflow'),
        installed: fs.existsSync(path.join(this.homeDir, '.cursor', 'skills', 'agentic-workflow'))
      },
      {
        name: 'Codex / OpenCode Skills',
        platform: 'codex',
        path: path.join(this.homeDir, '.codex', 'skills', 'agentic-workflow'),
        installed: fs.existsSync(path.join(this.homeDir, '.codex', 'skills', 'agentic-workflow'))
      },
      {
        name: 'Universal Agent Kernel',
        platform: 'agents',
        path: path.join(this.homeDir, '.agents', 'skills', 'agentic-workflow'),
        installed: fs.existsSync(path.join(this.homeDir, '.agents', 'skills', 'agentic-workflow'))
      }
    ];
    return targets;
  }

  public checkStatus(): InstallReport {
    const targets = this.getHostTargets();
    const installed = targets.filter(t => t.installed);
    const skipped = targets.filter(t => !t.installed);

    const binPaths = [
      path.join(this.homeDir, '.local', 'bin', 'agentic-workflow'),
      '/usr/local/bin/agentic-workflow'
    ];
    const linked = binPaths.filter(p => fs.existsSync(p));

    const systemDeps = [
      { name: 'bun', ...this.checkBinary('bun') },
      { name: 'node', ...this.checkBinary('node') },
      { name: 'python3', ...this.checkBinary('python3') },
      { name: 'git', ...this.checkBinary('git') },
      { name: 'brew', ...this.checkBinary('brew') }
    ];

    return {
      success: true,
      installedTargets: installed,
      skippedTargets: skipped,
      binLinked: linked,
      shellConfigured: [],
      systemDeps,
      messages: [
        `Installed in ${installed.length} agent host(s)`,
        `CLI binaries linked in ${linked.length} path(s)`
      ]
    };
  }

  public install(options: InstallOptions = {}): InstallReport {
    const messages: string[] = [];
    const installedTargets: InstallTarget[] = [];
    const skippedTargets: InstallTarget[] = [];
    const binLinked: string[] = [];
    const shellConfigured: string[] = [];

    const hostTargets = this.getHostTargets();
    const requestedPlatforms = options.targets || ['claude', 'gemini', 'cursor', 'codex', 'agents'];

    // 1. Copy or Symlink into host skills directories
    for (const target of hostTargets) {
      if (!requestedPlatforms.includes(target.platform)) {
        skippedTargets.push(target);
        continue;
      }

      const parentDir = path.dirname(target.path);
      try {
        if (!fs.existsSync(parentDir)) {
          fs.mkdirSync(parentDir, { recursive: true });
        }

        if (fs.existsSync(target.path)) {
          fs.rmSync(target.path, { recursive: true, force: true });
        }

        // Copy files excluding node_modules / .git
        fs.cpSync(this.projectDir, target.path, {
          recursive: true,
          filter: (src) => {
            const rel = path.relative(this.projectDir, src);
            if (rel.startsWith('node_modules') || rel.startsWith('.git') || rel.startsWith('__pycache__')) {
              return false;
            }
            return true;
          }
        });

        target.installed = true;
        installedTargets.push(target);
        messages.push(`✓ Installed skill to ${target.name} (${target.path})`);
      } catch (err: any) {
        messages.push(`! Failed installing to ${target.name}: ${err.message}`);
        skippedTargets.push(target);
      }
    }

    // 2. Global CLI Binary Symlink
    const cliSource = path.join(this.projectDir, 'bin', 'cli.js');
    if (fs.existsSync(cliSource)) {
      try {
        fs.chmodSync(cliSource, 0o755);
      } catch {
        // Ignore
      }

      const localBinDir = path.join(this.homeDir, '.local', 'bin');
      if (!fs.existsSync(localBinDir)) {
        fs.mkdirSync(localBinDir, { recursive: true });
      }

      const localBinTarget = path.join(localBinDir, 'agentic-workflow');
      try {
        if (fs.existsSync(localBinTarget)) {
          fs.unlinkSync(localBinTarget);
        }
        fs.symlinkSync(cliSource, localBinTarget);
        binLinked.push(localBinTarget);
        messages.push(`✓ Linked executable to ${localBinTarget}`);
      } catch (err: any) {
        messages.push(`! Failed linking ${localBinTarget}: ${err.message}`);
      }

      // Try /usr/local/bin if writable
      if (options.globalBin) {
        const usrBinTarget = '/usr/local/bin/agentic-workflow';
        try {
          if (fs.existsSync(usrBinTarget)) {
            fs.unlinkSync(usrBinTarget);
          }
          fs.symlinkSync(cliSource, usrBinTarget);
          binLinked.push(usrBinTarget);
          messages.push(`✓ Linked global executable to ${usrBinTarget}`);
        } catch {
          // May need sudo, skip gracefully
        }
      }
    }

    // 3. Update Shell RC if requested
    if (options.updateShellRc !== false) {
      const localBin = path.join(this.homeDir, '.local', 'bin');
      const exportLine = `export PATH="$PATH:${localBin}"`;
      const rcFiles = [
        path.join(this.homeDir, '.zshrc'),
        path.join(this.homeDir, '.bashrc'),
        path.join(this.homeDir, '.profile')
      ];

      for (const rc of rcFiles) {
        if (fs.existsSync(rc)) {
          try {
            const content = fs.readFileSync(rc, 'utf-8');
            if (!content.includes(localBin)) {
              fs.appendFileSync(rc, `\n# AgenticWorkflow CLI\n${exportLine}\n`);
              shellConfigured.push(rc);
              messages.push(`✓ Added PATH export to ${path.basename(rc)}`);
            }
          } catch {
            // Ignore
          }
        }
      }
    }

    // 4. System Dependencies Check
    const systemDeps = [
      { name: 'bun', ...this.checkBinary('bun') },
      { name: 'node', ...this.checkBinary('node') },
      { name: 'python3', ...this.checkBinary('python3') },
      { name: 'git', ...this.checkBinary('git') },
      { name: 'brew', ...this.checkBinary('brew') }
    ];

    return {
      success: installedTargets.length > 0 || binLinked.length > 0,
      installedTargets,
      skippedTargets,
      binLinked,
      shellConfigured,
      systemDeps,
      messages
    };
  }

  public generateShellCompletion(shell: 'zsh' | 'bash' | 'fish'): string {
    const commands = [
      'engine', 'autopilot', 'run', 'skills', 'guard', 'eval', 'traces',
      'hooks', 'integrations', 'fable', 'toon', 'init', 'validate', 'status',
      'test', 'update', 'install', 'refresh', 'doctor', 'health', 'deps',
      'notify', 'announcements', 'version', 'help'
    ].join(' ');

    if (shell === 'zsh') {
      return `#compdef agentic-workflow
_agentic_workflow() {
  local -a commands
  commands=(${commands})
  _describe 'command' commands
}
compdef _agentic_workflow agentic-workflow
`;
    } else if (shell === 'fish') {
      return `complete -c agentic-workflow -f -a "${commands}"`;
    } else {
      return `_agentic_workflow() {
  local cur="\${COMP_WORDS[COMP_CWORD]}"
  COMPREPLY=( $(compgen -W "${commands}" -- "$cur") )
}
complete -F _agentic_workflow agentic-workflow
`;
    }
  }
}

File: core/system/installer.py
"""
core/system/installer.py — Python Auto-Installer Engine for AgenticWorkflow
"""

import os
import shutil
import subprocess
from pathlib import Path
from typing import Dict, Any, List, Optional


class AutoInstaller:
    def __init__(self, project_dir: str = "."):
        self.project_dir = Path(project_dir).resolve()
        self.home_dir = Path.home()

    def _check_binary(self, cmd: str) -> Dict[str, Any]:
        try:
            res = subprocess.run(f"{cmd} --version", shell=True, capture_output=True, text=True)
            return {"available": res.returncode == 0, "version": res.stdout.strip().split("\n")[0]}
        except Exception:
            return {"available": False}

    def get_host_targets(self) -> List[Dict[str, Any]]:
        return [
            {
                "name": "Claude Code Skills",
                "platform": "claude",
                "path": str(self.home_dir / ".claude" / "skills" / "agentic-workflow"),
                "installed": (self.home_dir / ".claude" / "skills" / "agentic-workflow").exists()
            },
            {
                "name": "Gemini CLI / Antigravity Skills",
                "platform": "gemini",
                "path": str(self.home_dir / ".gemini" / "config" / "skills" / "agentic-workflow"),
                "installed": (self.home_dir / ".gemini" / "config" / "skills" / "agentic-workflow").exists()
            },
            {
                "name": "Cursor IDE Skills",
                "platform": "cursor",
                "path": str(self.home_dir / ".cursor" / "skills" / "agentic-workflow"),
                "installed": (self.home_dir / ".cursor" / "skills" / "agentic-workflow").exists()
            },
            {
                "name": "Codex / OpenCode Skills",
                "platform": "codex",
                "path": str(self.home_dir / ".codex" / "skills" / "agentic-workflow"),
                "installed": (self.home_dir / ".codex" / "skills" / "agentic-workflow").exists()
            },
            {
                "name": "Universal Agent Kernel",
                "platform": "agents",
                "path": str(self.home_dir / ".agents" / "skills" / "agentic-workflow"),
                "installed": (self.home_dir / ".agents" / "skills" / "agentic-workflow").exists()
            }
        ]

    def check_status(self) -> Dict[str, Any]:
        targets = self.get_host_targets()
        installed = [t for t in targets if t["installed"]]
        skipped = [t for t in targets if not t["installed"]]

        bin_paths = [
            str(self.home_dir / ".local" / "bin" / "agentic-workflow"),
            "/usr/local/bin/agentic-workflow"
        ]
        linked = [p for p in bin_paths if os.path.exists(p)]

        deps = [
            {"name": "bun", **self._check_binary("bun")},
            {"name": "node", **self._check_binary("node")},
            {"name": "python3", **self._check_binary("python3")},
            {"name": "git", **self._check_binary("git")}
        ]

        return {
            "success": True,
            "installed_targets": installed,
            "skipped_targets": skipped,
            "bin_linked": linked,
            "system_deps": deps
        }

    def install(self, platforms: Optional[List[str]] = None, global_bin: bool = True) -> Dict[str, Any]:
        requested = platforms or ["claude", "gemini", "cursor", "codex", "agents"]
        messages = []
        installed_targets = []
        bin_linked = []

        targets = self.get_host_targets()
        for t in targets:
            if t["platform"] not in requested:
                continue

            target_path = Path(t["path"])
            target_path.parent.mkdir(parents=True, exist_ok=True)
            if target_path.exists():
                shutil.rmtree(target_path, ignore_errors=True)

            try:
                shutil.copytree(
                    self.project_dir,
                    target_path,
                    ignore=shutil.ignore_patterns("node_modules", ".git", "__pycache__", ".pytest_cache")
                )
                t["installed"] = True
                installed_targets.append(t)
                messages.append(f"✓ Installed skill to {t['name']}")
            except Exception as e:
                messages.append(f"! Failed installing to {t['name']}: {e}")

        # CLI symlink
        cli_src = self.project_dir / "bin" / "cli.js"
        if cli_src.exists():
            try:
                os.chmod(cli_src, 0o755)
            except Exception:
                _chmod_ok = False

            local_bin = self.home_dir / ".local" / "bin"
            local_bin.mkdir(parents=True, exist_ok=True)
            link_target = local_bin / "agentic-workflow"
            try:
                if link_target.exists() or link_target.is_symlink():
                    link_target.unlink()
                link_target.symlink_to(cli_src)
                bin_linked.append(str(link_target))
                messages.append(f"✓ Linked executable to {link_target}")
            except Exception as e:
                messages.append(f"! Failed linking CLI: {e}")

        return {
            "success": len(installed_targets) > 0 or len(bin_linked) > 0,
            "installed_targets": installed_targets,
            "bin_linked": bin_linked,
            "messages": messages
        }
